sw/host/provisioning/ft_lib/src/lib.rs

```diff
@@ -247,27 +249,103 @@ pub fn run_ft_personalize(
     log::info!("UDS Cert: {}", hex::encode(uds_cert_bytes.clone()));
+    let _ = parse_certificate(&uds_cert_bytes)?;
     log::info!("CDI_0 Cert: {}", hex::encode(cdi_0_cert_bytes.clone()));
+    let _ = parse_certificate(&cdi_0_cert_bytes)?;
     log::info!("CDI_1 Cert: {}", hex::encode(cdi_1_cert_bytes.clone()));
-    let _uds_cert = parse_certificate(&uds_cert_bytes)?;
-    let _cdi_0_cert = parse_certificate(&cdi_0_cert_bytes)?;
-    let _cdi_1_cert = parse_certificate(&cdi_1_cert_bytes)?;
+    let _ = parse_certificate(&cdi_1_cert_bytes)?;
+    log::info!("TPM EK Cert: {}", hex::encode(tpm_ek_cert_bytes.clone()));
+    let _ = parse_certificate(&tpm_ek_cert_bytes)?;
+    log::info!("TPM CEK Cert: {}", hex::encode(tpm_cek_cert_bytes.clone()));
+    let _ = parse_certificate(&tpm_cek_cert_bytes)?;
+    log::info!("TPM CIK Cert: {}", hex::encode(tpm_cik_cert_bytes.clone()));
+    let _ = parse_certificate(&tpm_cik_cert_bytes)?;
+
+    // Hash all certificates' contents.
+    let mut hasher = Sha256::new();
+    let all_certs: [&Vec<u8>; 6] = [
+        &uds_cert_bytes,
+        &cdi_0_cert_bytes,
+        &cdi_1_cert_bytes,
+        &tpm_ek_cert_bytes,
+        &tpm_cek_cert_bytes,
+        &tpm_cik_cert_bytes,
+    ];
+    for cert in all_certs {
+        // Do not hash trailing bytes present in the certificate vectors, use
+        let size = u16::from_be_bytes([cert[2], cert[3]]) as usize;
+        let mut cert_clone = cert.clone();
+        if cert_clone.len() > size {
+            cert_clone.truncate(size);
+        }
+        hasher.update(cert_clone);
+        log::info!("Running hash {:?}", hex::encode(hasher.clone().finalize()));
+    }
 
     // Send endorsed certificates back to the device.
     let endorsed_certs = ManufEndorsedCerts {
         uds_certificate: uds_cert_bytes.clone().into_iter().collect(),
         uds_certificate_size: uds_cert_bytes.len(),
+        tpm_ek_certificate: tpm_ek_cert_bytes.clone().into_iter().collect(),
+        tpm_ek_certificate_size: tpm_ek_cert_bytes.len(),
+        tpm_cek_certificate: tpm_cek_cert_bytes.clone().into_iter().collect(),
+        tpm_cek_certificate_size: tpm_cek_cert_bytes.len(),
+        tpm_cik_certificate: tpm_cik_cert_bytes.clone().into_iter().collect(),
+        tpm_cik_certificate_size: tpm_cik_cert_bytes.len(),
     };
-    let _ = UartConsole::wait_for(&*uart, r"Importing DICE UDS certificate ...", timeout)?;
+    let _ = UartConsole::wait_for(&*uart, r"Importing endorsed certificates ...", timeout)?;
     endorsed_certs.send(&*uart)?;
-    let _ = UartConsole::wait_for(&*uart, r"Imported DICE UDS certificate.", timeout)?;
-
-    Ok(())
+    let _ = UartConsole::wait_for(&*uart, r"Final hash:", timeout)?;
+    let received_hash = SerdesSha256Hash::recv(&*uart, timeout, false)?;
+    let _ = UartConsole::wait_for(&*uart, r"Finished importing certificates.", timeout)?;
+
+    // Compare host and device generated certificates' hash values.
+    if received_hash
```

Comment:
I think for debugging it would be useful to construct the two hashs in array and then compare them. Also zipping them seems a bit overkill. Shouldn't the follow just work?
```rust
let actual_hash = hash.finalize();
if data.as_bytes() == actual_hash.as_slice() { // I think the as_slice might be necessary but I didn't check the types
}
```

Reply:
after changing the hashing function on the device the generated hash endianness needs to change, which is easier when using this zipping pattern. Also, if hashes don't match the hash received from the device is printed out later.